### ISSUE TITLE
`map` forgets its shrinker

### ISSUE THREAD

**Issue by Chattered** [kotlintest/kotlintest#472]
For example:

```
Gen.string().shrinker != null
Gen.string().map { it }.shrinker() == null
```

This means we don't get a free shrinker when combining Gens. Moreover, it is surprising behaviour for users since it violates the functor law for map.

If it cannot be implemented without breaking covariance, then I think shrinker() should be removed from the Gen interface, and Gen should go back to being covariant.

**Comment by Kerooker** [comment]
You're correct. I believe the implementation for Gen at the moment doesn't consider mapping in a correct way. I'll see what can be done about it.

Thanks for reporting!

**Comment by Kerooker** [comment]
Ok, so here's the thing:
If you're mapping a type for the same type (for example, mapping Int to Int), the shrinker should be the same. However. what if I map one value to another?

For example:

`Gen<Int>.map { it.toString }`

It's impossible to create a shrinker based on this.

What I'll do is return the original shrinker if the type is the same, and return a shrinker if the time is known. Else it will remain null

**Comment by Chattered** [comment]
Unfortunately, it really weakens composability. When you present a monad interface, you expect users to be able to build up complex Gens using map, flatMap, and the various "bind" functions in the Gen interface. These compositions should always be valid according to the monad interface. 

We're using your library in our company (thanks for writing it, btw!) to do what we consider to be obvious things like:

```
Gen.bind(Gen.int(), Gen.string(), { n,str -> Pair(n,str) })
```

and we expect this to be build a composite generator. But we're surprised to lose the Shrinker.

To be pedantic, your proposal still breaks the functor laws, namely the composition law:

Gen<Int>.map { it.toString }.map { Integer.parseInt(it) }.shrinker() == null
Gen<Int>.map { Integer.parseInt(it.toString) }.shrinker() != null
```

My suggestion is still to remove Shrinker from the interface. Is it possible, instead, to pass a Shrinker to functions like "forall", once the final type is known? Perhaps this could be done implicitly, on the assumption that every type determines its shrinker, rather than letting shrinkers vary by generator.

**Comment by sksamuel** [comment]
To be fair, just because there's a flatMap method, it doesn't mean Gen claims to be a Monad. Similarly for a Functor and map. 

I like your suggestion of passing the shrinker to forAll/assertAll. 
Another alternative is requiring the user to supply a fn to map the shrinker as well as the generated values.

Although in the case of bind, there's no real reason why the shrinker can't be created automatically. I think the reason I never did this is because of the "edge cases" that our shrinkers provide, and the combinatorial explosion.

**Comment by Kerooker** [comment]
`Gen.int().map { it.toString }.map { Integer.parseInt(it) }.shrinker() == null`
In this example, it wouldn't violate that law, as Integer is a known type and there would be a shrinker for it. The same as the second example.

**Comment by Chattered** [comment]
```
Gen.int().map { it.toString }.map { Integer.parseInt(it) }.shrinker() == null
```

would violate the law. The first map changes the type from Integer to String, so the Shrinker gets chucked. The second map doesn't restore the Shrinker. So you lose it at the end. However, if you fuse the maps together, you're mapping from Integer to Integer, so your typecheck would preserve the Shrinker.

**Comment by Kerooker** [comment]
After the second map, if the shrinker is null and I know that I have a shrinker for it's type (Int), so I would be able to provide the shrinker for it. Would you expect the second map to have an Int shrinker?

**Comment by Chattered** [comment]
Ah, I missed the bit about returning a shrinker if the type is known. That seems like a possible tradeoff, if we can easily (and are encouraged to) extend the known types.

I still think there might be a better way to do this. And it seems like a bad smell to me that a type supporting map is invariant.

**Comment by sksamuel** [comment]
Also as an addition, I don't think having generators detected by type along is good enough. For example, even and odd generators require different shrinkers but both are Gen<Int>.

**Comment by Chattered** [comment]
FWIW: I've found myself writing specialised flatMaps that drop the edge cases to deal with the combinatorial explosion. It's a real shame, because on first look, the idea of edge cases seems like a really cool idea.

Looking at Haskell, shrink is part of the Arbitrary typeclass but not the Gen typeclass, which might indicate that these things need to be kept separate, perhaps by supplying the shrinker independently of the composition of Gens as I suggested above. Composition of shrinkers might then need to be part of a separate interface.

**Comment by sksamuel** [comment]
I like that idea, but I'd like to think of a way that a particular gen could supply a shrinker as well. For example, if I want to use a Gen for even numbers, I'd have to manually supply the shrinker, going by type would be not be enough. How does Haskell solve that, or does it just require you to provide the right one.

**Comment by Chattered** [comment]
I'd need a concrete example. Cases where I've needed subsets of integers like that, I've normally gotten away with setting things up right at the test-site. So if property p must be true of all evens, I just say:

test n = p (2 * n)

and I'm done.

**Comment by sksamuel** [comment]
I just mean if you have a `Gen<Int>` that returns evens, then lets say your test failed for `4`, then a standard `Shrinker<Int>` is going to return 3, 2, 1 which aren't compatible with the generator.

**Comment by Chattered** [comment]
I mostly write quickcheck tests in Haskell, and I haven't needed such a generator. I'd want to see a concrete example of a case where I need to generate even numbers that isn't better solved by multiplying an integer by 2 at the test-site.

**Comment by sksamuel** [comment]
I suppose that example is easily created from a normal gen int. What about a prime number generator, or an ascii string generator (subset of all strings),

**Comment by Chattered** [comment]
I think if you really need to work with a subtype like that, you want to define a newtype wrapper and hide the data constructor, so you can enforce the fact that your string is ascii in the type system. You can then define an Arbitrary instance for the newtype which does the appropriate shrinking.

In Kotlin, that would be a data class with a private constructor. It doesn't get the benefit you get in Haskell (and Scala's value classes), which guarantee that the newtype doesn't take any more space than the underlying type, but you could get the same behaviour.

**Comment by sksamuel** [comment]
Yes that's one solution I suppose.

If I want to write a custom generator for strings, then I can either lose shrinker support, wrap the type in a data class, or have an overloaded forAll that accepts a specific shrinker. Out of those, I don't think there's much to choose between the last two, in that they're both not ideal, but I do think they're probably both better than the current situation.

**Comment by Kerooker** [comment]
@Chattered Maybe one can use the new inline class for that? It will take exactly the same space as the underlying value, and has type safety

**Comment by Chattered** [comment]
@Kerooker Nice! That sounds exactly like a newtype/value-class.

**Comment by Chattered** [comment]
@Kerooker Nice! That sounds exactly like a newtype/value-class.

**Comment by Chattered** [comment]
Though can inline classes be used in the proposed case? When picking out a shrinker based on a known type, you'd have to be able to do it statically, since inline classes will lose their wrapper at runtime and just appear as the underlying type, right?

**Comment by sksamuel** [comment]
I think that's right Phil. The underlying type is lost.

**Comment by LeoColman** [comment]
Closed as per #1135
